onedal/datatypes/dlpack/data_conversion.cpp

```diff
@@ -0,0 +1,183 @@
+/*******************************************************************************
+* Copyright Contributors to the oneDAL project
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/table/homogen.hpp"
+#include "oneapi/dal/table/detail/homogen_utils.hpp"
+
+#include "onedal/datatypes/dlpack/data_conversion.hpp"
+
+#ifdef ONEDAL_DATA_PARALLEL
+#include "onedal/common/sycl_interfaces.hpp"
+#endif // ONEDAL_DATA_PARALLEL
+
+namespace oneapi::dal::python::dlpack {
+
+template <typename T, typename managed_t>
+inline dal::homogen_table convert_to_homogen_impl(managed_t* dlm_tensor,
+                                                  bool readonly,
+                                                  py::object q_obj) {
+    dal::homogen_table res{};
+
+    DLTensor tensor = dlm_tensor->dl_tensor;
+
+    // if a nullptr, return an empty.
+    if (!tensor.data) {
+        return res;
+    }
+
+    // Get pointer to the data following dlpack.h conventions.
+    // use static cast because data is known to be void*, and reintrepret_cast
+    // since we know that T* is of a set specific types that we will guarantee
+    // for the compiler.
+    const auto* const ptr =
+        reinterpret_cast<const T*>(static_cast<char*>(tensor.data) + tensor.byte_offset);
```

Comment:
```suggestion
    auto raw_data = static_cast<char*>(tensor.data);
    const auto ptr = reinterpret_cast<const T*>(raw_data + tensor.byte_offset);
```

Not sure this is the best way, but I would consider a few things.

- Don't like the double `static_cast` in a single line
- Do we even need to convert to `char*` when all we do is adding an offset and casting again? The `operator+()` would produce the same target address, regardless of the base being `char*` or `void*`, wouldn't it?
- Is `const auto* const` really what you want? We don't often declare pointers themselves `const`, so I'm just wondering.

Reply:
For point 2) https://stackoverflow.com/a/3524270
Otherwise good call, will separate the casts, and look at the constness more closely

Reply:
For ```const auto* const```, dlpack data is immutable, and we don't want to move its location, so it a const pointer to const data. Luckily dal tables are okay with that out of the box.